backend/app/services/knowledge/structured/duckdb_storage.py

```
@@ -0,0 +1,334 @@
+
+    Each knowledge base has its own isolated in-memory DuckDB instance.
+    Data is loaded from source files and cached in memory for fast querying.
+
+    Thread Safety:
+        DuckDB connections are not thread-safe. Each thread should use its own
+        connection. This class creates new connections per operation for safety.
+    """
+
+    # Class-level storage for DuckDB data (table_name -> DataFrame)
+    # We store DataFrames instead of connections for thread safety
+    _data_store: Dict[str, pd.DataFrame] = {}
```

Ruff (0.15.0)

[warning] Mutable class attributes should be annotated with `typing.ClassVar`

(RUF012)